src/bijectors/planar_layer.jl

```diff
@@ -15,73 +15,93 @@ using Roots # for inverse
 
 # TODO: add docstring
 
-# FIXME: using `TrackedArray` for the parameters, we end up with
-# nested tracked structures; don't want this.
-mutable struct PlanarLayer{T1,T2} <: Bijector{1}
+mutable struct PlanarLayer{T1 <: AbstractVector{<:Real}, T2 <: Real} <: Bijector{1}
     w::T1
     u::T1
     b::T2
 end
 
 function get_u_hat(u, w)
     # To preserve invertibility
-    # From A.1
-    return u .+ (planar_flow_m.(w' * u) .- w' * u) .* w ./ sum(abs2, w)
+    x = w' * u
+    return u .+ (planar_flow_m(x) - x) .* w ./ sum(abs2, w)   # from A.1
 end
 
 function PlanarLayer(dims::Int, container=Array)
-    w = container(randn(dims, 1))
-    u = container(randn(dims, 1))
-    b = container(randn(1))
+    w = container(randn(dims))
+    u = container(randn(dims))
+    b = randn()
     return PlanarLayer(w, u, b)
 end
 
 planar_flow_m(x) = -1 + softplus(x)   # for planar flow from A.1
-dtanh(x) = 1 - (tanh(x)) ^ 2         # for planar flow
-ψ(z, w, b) = dtanh.(w' * z .+ b) .* w    # for planar flow from eq(11)
+ψ(z, w, b) = (1 .- tanh.(w' * z .+ b).^2) .* w    # for planar flow from eq(11)
 
 # An internal version of transform that returns intermediate variables
-function _transform(flow::PlanarLayer, z)
-    u_hat = get_u_hat(flow.u, flow.w)
-    transformed = z + u_hat * tanh.(flow.w' * z .+ flow.b) # from eq(10)
-    return (transformed=transformed, u_hat=u_hat)
+function _transform(flow::PlanarLayer, z::AbstractVecOrMat)
+    _planar_transform(flow.u, flow.w, flow.b, z)
+end
+function _planar_transform(u, w, b, z)
+    u_hat = get_u_hat(u, w)
+    transformed = z .+ u_hat .* tanh.(w' * z .+ b) # from eq(10)
+    return (transformed = transformed, u_hat = u_hat)
 end
 
 (b::PlanarLayer)(z) = _transform(b, z).transformed
 
-function _forward(flow::PlanarLayer, z)
+function forward(flow::PlanarLayer, z::AbstractVecOrMat)
     transformed, u_hat = _transform(flow, z)
     # Compute log_det_jacobian
-    psi = ψ(z, flow.w, flow.b)
-    log_det_jacobian = log.(abs.(1.0 .+ psi' * u_hat))          # from eq(12)
-    return (rv=transformed, logabsdetjac=vec(log_det_jacobian)) # from eq(10)
+    psi = ψ(z, flow.w, flow.b) .+ zero(eltype(u_hat))
+    if psi isa AbstractVector
+        T = eltype(psi)
+    else
+        T = typeof(vec(psi))
+    end
+    log_det_jacobian::T = log.(abs.(1.0 .+ psi' * u_hat)) # from eq(12)
```

Comment:
Why is all this needed? Seems strange that the type of `log_det_jacobian` isn't inferrable, no?

Reply:
Whenever `TrackedArray` and broadcasting are involved, it seems that type inference jumps out of the window.
```julia
julia> using Tracker

julia> f(x) = log.(abs.(x))
f (generic function with 1 method)

julia> @code_warntype f(param(rand(3)))
Variables
  #self#::Core.Compiler.Const(f, false)
  x::TrackedArray{…,Array{Float64,1}}

Body::Any
1 ─ %1 = Base.broadcasted(Main.abs, x)::Base.Broadcast.Broadcasted{Tracker.TrackedStyle,Nothing,typeof(abs),Tuple{TrackedArray{…,Array{Float64,1}}}}
│   %2 = Base.broadcasted(Main.log, %1)::Base.Broadcast.Broadcasted{Tracker.TrackedStyle,Nothing,typeof(log),Tuple{Base.Broadcast.Broadcasted{Tracker.TrackedStyle,Nothing,typeof(abs),Tuple{TrackedArray{…,Array{Float64,1}}}}}}
│   %3 = Base.materialize(%2)::Any
└──      return %3
```